--- parse.y	13 Oct 2004 19:46:40 -0000	1.1.1.1
+++ parse.y	3 Dec 2004 01:25:53 -0000
@@ -267,6 +267,7 @@
 %type <num>  f_norm_arg f_arg
 %token tUPLUS 		/* unary+ */
 %token tUMINUS 		/* unary- */
+%token tUBS             /* unary\ */
 %token tPOW		/* ** */
 %token tCMP  		/* <=> */
 %token tEQ  		/* == */
@@ -1797,6 +1798,14 @@
 		    {
 			$$ = new_call($1, $3, 0);
 		    }
+		| primary_value '\\' operation2
+		    {
+		        $$ = NEW_CALL($1, rb_intern("method"), NEW_LIST(NEW_LIT(ID2SYM($3))));
+		    }
+		| tUBS operation2
+		    {
+		        $$ = NEW_FCALL(rb_intern("method"), NEW_LIST(NEW_LIT(ID2SYM($2))));
+		    }
 		| kSUPER paren_args
 		    {
 			$$ = new_super($2);
@@ -4104,6 +4113,11 @@
 	    goto retry; /* skip \\n */
 	}
 	pushback(c);
+	if (lex_state == EXPR_BEG || lex_state == EXPR_MID || space_seen) {
+	  lex_state = EXPR_DOT;
+	  return tUBS;
+	}
+	lex_state = EXPR_DOT;
 	return '\\';
 
       case '%':