--- a/load.c  2010-10-23 05:36:38.000000000 -0400
+++ b/patchload.c 2011-06-05 08:58:00.000000000 -0400
@@ -40,14 +40,6 @@
     VALUE ary;
     long i;

-    for (i = 0; i < RARRAY_LEN(load_path); ++i) {
- VALUE str = rb_check_string_type(RARRAY_PTR(load_path)[i]);
- if (NIL_P(str) || !rb_is_absolute_path(RSTRING_PTR(str)))
-     goto relative_path_found;
-    }
-    return load_path;
-
-  relative_path_found:
     ary = rb_ary_new2(RARRAY_LEN(load_path));
     for (i = 0; i < RARRAY_LEN(load_path); ++i) {
  VALUE path = rb_file_expand_path(RARRAY_PTR(load_path)[i], Qnil);
@@ -81,16 +73,27 @@
        int type, VALUE load_path)
 {
     long i;
+    long plen;
+    const char *e;

+    if(vlen < len) return 0;
+    if (!strncmp(name+(vlen-len),feature,len)){
+ plen = vlen - len - 1;
+    } else {
+ for (e = name + vlen; name != e && *e != '.' && *e != '/'; --e);
+ if (*e!='.' ||
+     e-name < len ||
+     strncmp(e-len,feature,len) )
+     return 0;
+ plen = e - name - len - 1;
+    }
     for (i = 0; i < RARRAY_LEN(load_path); ++i) {
  VALUE p = RARRAY_PTR(load_path)[i];
  const char *s = StringValuePtr(p);
  long n = RSTRING_LEN(p);

- if (vlen < n + len + 1) continue;
+ if (n != plen ) continue;
  if (n && (strncmp(name, s, n) || name[n] != '/')) continue;
- if (strncmp(name + n + 1, feature, len)) continue;
- if (name[n+len+1] && name[n+len+1] != '.') continue;
  switch (type) {
    case 's':
      if (IS_DLEXT(&name[n+len+1])) return p;
